src/ripple/basics/Number.h

```diff
@@ -0,0 +1,340 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_NUMBER_H_INCLUDED
+#define RIPPLE_BASICS_NUMBER_H_INCLUDED
+
+#include <ripple/basics/IOUAmount.h>
+#include <ripple/basics/XRPAmount.h>
+#include <cstdint>
+#include <ostream>
+#include <string>
+
+namespace ripple {
+
+class Number;
+
+std::string
+to_string(Number const& amount);
+
+class Number
+{
+    using rep = std::int64_t;
+    rep mantissa_{0};
+    int exponent_{-2'147'483'648};
+
+public:
+    struct unchecked
+    {
+        explicit unchecked() = default;
+    };
+
+    explicit Number() = default;
+
+    Number(rep mantissa);
+    explicit Number(rep mantissa, int exponent);
```

Comment:
I'd rather not move `normalize()` into the header because of its size.  The compiler _probably_ won't inline it, which is good, but then the linker has to spend time eliminating duplicate outlined copies, which is bad.  My policy is to not risk the code bloat or the extra linker work unless there is a benefit that outweighs these costs.

However sometimes I _do_ want to have `constexpr` values.  Thus I introduced:

```c++
explicit constexpr Number(rep mantissa, int exponent, unchecked) noexcept;
```
and 
```c++
constexpr bool  isnormal() const noexcept;
```
Here is an example use: https://github.com/ripple/rippled/pull/4145/commits/9968d47146f07b164ae79982d597c461c8a60bf5#diff-6c9395b2018418d46bb1c2584aa296d79ab620de7615538c6734648ca6eb71aeR357-R362